### PR TITLE
[WIP] Added skeleton of python dispatch key

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56718 [WIP] Added skeleton of python dispatch key**
* #57426 [nnc] Added NNC lowerings for t/transpose/permute/expand + other cleaning

### CODE REVIEW

**Comment by Chillee** [comment]
This a ghstacked version of https://github.com/pytorch/pytorch/pull/56137.

**Comment by Chillee** [comment]
Got an example of tracing through (mostly) a simple linear model, and compiling the whole thing with NNC.

NNC is about 3x faster in this whole example, although if you ignore the optimizer updates, NNC is about 10x faster than eager.

image

File: torch/csrc/fx/fx_init.cpp [hunk 1]
```diff
@@ -137,6 +159,138 @@ void initFx(PyObject* module) {
...
+	if (!tuple) throw std::runtime_error("Unable to allocate memory for Python tuple");
+	for (unsigned int i = 0; i < data.size(); i++) {
+		PyObject *num = converter(data[i]).ptr();
+		if (!num) {
+			Py_DECREF(tuple);
+			throw std::runtime_error("Unable to allocate memory for Python tuple");
+		}
+    Py_INCREF(num); // todo: dunno?? Need it to fix segfaults, but probably not right
+		PyTuple_SET_ITEM(tuple, i, num);
+	}
```
Comment (jamesr66a) [inline torch/csrc/fx/fx_init.cpp:187]: don't use tabs

Comment (Chillee) [reply torch/csrc/fx/fx_init.cpp:187]: Oh, this used to have a [WIP][not for review] on it, before I switched to ghstack.

**Pull request closed without merging**